### PR TITLE
Evaluate DisplayNameGenerator.ReplaceUnderscores

### PR DESCRIPTION
**Pull request by thmuch**
I like to use JUnit 5's feature for automatic display name generation, especially the strategy for replacing underscores in test method names by spaces. This way, I can phrase the test names like a human readable sentence – without having to use ```@DisplayName``` all the time.

Unfortunately, this doesn't work with ArchUnit yet. On the one hand, ArchUnitTestDescriptor isn't derived (or does not use) the specialised Jupiter descriptor classes. But refactoring this might be too much of a change. And on the other hand, JUnit's DisplayNameGenerator API isn't really suited for ArchUnit's way of dealing with tests, especially the field based tests.

So I tried a really simple approach: Evaluating only the one DisplayNameGenerator strategy that makes sense most IMHO for ArchUnit (ReplaceUnderscores) with an implementation not (syntactically) derived from JUnit's implementation of the DisplayNameGenerator API.

With the changes in this PR it's possible to write

```java
@AnalyzeClasses(packages = "some.dummy.package")
@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
public class SomeTestClass { ... }
```

for field and method based ArchUnit tests.

Feel free to reject this PR if it is too little benefit for too much change. Or if you think this is crude use of a public API. But I tried to be as little invasive as possible, while still solving my main use case.

### CODE REVIEW

**Comment by thmuch** [comment]
Hm... thinking about it... Is it ok that we'd required all JUnit 5 users to use at least JUnit 5.4 (when DisplayNameGenerator became an experimental API) or JUnit 5.7 (when die API was declared stable)? Because now the Jupiter API is an implementation dependency in ArchUnit's JUnit 5 engine artifact.

**Comment by thmuch** [comment]
Another option for implementing this feature would be to create synthetic Method objects for the test fields (and strip the JavaClasses param from test methods). With that approach, we probably could (re)use all existing DisplayNameGenerator implementations. But that would feel somewhat hacky, too.
Maybe we keep this option in mind if the simple implementation isn't good enough any more.

**Comment by codecholeric** [comment]
Thanks a lot for the initiative! :smiley: You're right that the test method names themselves aren't super pretty. 
However, as you noted, the philosophy so far was to be completely independent of the Jupiter API and only depend on the JUnit platform :thinking: 
I would also like to keep it that way, because it is one less dependency to care about and more important the mental model of ArchUnit's tests just doesn't completely match the Jupiter model (which is why I went a separate road back then, instead of integrating it into the Jupiter engine as extension, etc.)
That being said we could still think about a custom implementation within the `ArchUnitTestEngine`, but it would need to be an ArchUnit API then as well I think, and completely independent of the Jupiter specifics :thinking: 
What do you think about it? We could also ponder if we shouldn't just always replace underscores by spaces within display names of the tests and not even make it customizable :thinking: Because I can't imagine anybody would complain about that, i.e. not seeing snake case but nicely formatted test display names. But you never know :man_shrugging:

**Comment by thmuch** [comment]
+1 for keeping ArchUnit as independant from Jupiter as possible, especially if it was a deliberate decision (although you have different modules for JUnit 4 and 5 already; and users of ArchUnit and JUnit 5 probably will already have a dependency to the Jupiter API already)

If we make replacing underscores in test names an ArchUnit custom feature, maybe a property would be enough? Like ```displayName.replaceUnderscores```? I'd make ```false``` the default, though, as using this feature should be a deliberate decision IMHO.

**Comment by thmuch** [comment]
Now underscores in test names are replaced by spaces if the property ```displayName.replaceUnderscores``` is set to ```true```.

If you think this is a good approach, I could try to integrate this kind of display name generation in archunit-junit/junit4. Where to place ```DisplayNameResolver``` best, in the "base" package of the "archunit" module? And make it ```@Internal``` then?

File: archunit-junit/junit5/engine/src/test/java/com/tngtech/archunit/junit/ArchUnitTestEngineTest.java [hunk 1]
```diff
@@ -1122,6 +1116,61 @@ void with_reason() {
         }
     }
 
+    @Nested
+    class GeneratesDisplayName {
+        @Test
+        void for_field_by_replacing_underscores_with_blanks_if_property_is_set_to_true() {
+            EngineDiscoveryTestRequest discoveryRequest = new EngineDiscoveryTestRequest().withClass(SimpleRuleField.class);
+
+            ArchConfiguration.get().setProperty(DisplayNameResolver.DISPLAYNAME_REPLACE_UNDERSCORES_PROPERTY_NAME, "true");
```
Comment (codecholeric) [inline archunit-junit/junit5/engine/src/test/java/com/tngtech/archunit/junit/ArchUnitTestEngineTest.java:1125]: If we do this we should also have some

```
@AfterEach
void resetConfiguration() {
    ArchConfiguration.get().reset();
}
```

Otherwise other tests might start to behave surprisingly / flaky :wink:

File: archunit-junit/junit5/engine/src/main/java/com/tngtech/archunit/junit/DisplayNameResolver.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    static String determineDisplayName(String elementName) {
+        if (replaceUnderscores()) {
+            return underscoresReplacedBySpaces(elementName);
+        } else {
+            return elementName;
+        }
+    }
+
+    private static String underscoresReplacedBySpaces(String elementName) {
```
Comment (codecholeric) [inline archunit-junit/junit5/engine/src/main/java/com/tngtech/archunit/junit/DisplayNameResolver.java:32]: Should we make the property also `replaceUnderscoresBySpaces`? I like this method name, because it says exactly what it does. The property name just says `replaceUnderscores`, but not with what :thinking:

File: archunit-junit/junit5/engine/src/main/java/com/tngtech/archunit/junit/DisplayNameResolver.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.junit;
+
+import com.tngtech.archunit.ArchConfiguration;
+
+final class DisplayNameResolver {
+
+    static final String DISPLAYNAME_REPLACE_UNDERSCORES_PROPERTY_NAME = "displayName.replaceUnderscores";
```
Comment (codecholeric) [inline archunit-junit/junit5/engine/src/main/java/com/tngtech/archunit/junit/DisplayNameResolver.java:22]: I would prefix the property name by `junit.` if it is for both JUnit versions and `junit5.` if it is only for JUnit 5. Because it does not influence the core of ArchUnit like most other properties.

**Review by codecholeric — COMMENTED** [review]
I like this approach! :+1: Added some minor comments, what do you think about them? Also I think it would be nice if it would work consistently for JUnit 4 and 5, but up to you if you want to invest the effort for now :wink: 
If you would add the support also to JUnit 4, then at the moment those modules share a common package anyway, so `DisplayNameResolver` still wouldn't have to be public (this will be a problem sooner or later, since we're violating the Java 9 modules contract here, but that is to be solved at a different time and there is more to do then :wink:)
In other words there already is a common module also containing the package `com.tngtech.archunit.junit`.

File: archunit-junit/junit5/engine/src/main/java/com/tngtech/archunit/junit/DisplayNameResolver.java (hunk 3 above)
Comment (thmuch) [reply archunit-junit/junit5/engine/src/main/java/com/tngtech/archunit/junit/DisplayNameResolver.java:22]: Thanks for your feedback! All suggestions seem reasonable.
I'll amend the pull request next week (hopefully), including JUnit 4 support.

**Comment by thmuch** [comment]
@codecholeric: This PR should be feature complete, finally. Unless you have some feedback, of course 🙃
